### ISSUE TITLE
2021.12: If the user has client tools for both CentOS8 and EL8 (Rocky/Alma/etc), the schema upgrade fails

### ISSUE THREAD

**Issue by juliogonzalez** [uyuni-project/uyuni#4557]
# Additional Information

_The following information is very important in order to help us help you. Omission of the following details can cause delays or can receive no attention at all._

## Version of Uyuni Server and Proxy (if used)
2021.12

## Details about issue
If the user has client tools for both CentOS8 and EL8 (Rocky/Alma/etc), the schema upgrade fails.

The problem is:
https://github.com/uyuni-project/uyuni/blob/master/schema/spacewalk/upgrade/susemanager-schema-4.3.3-to-susemanager-schema-4.3.4/001-centos8-eol.sql

If the EL8 client tools are already added, then this fails.

The fix is, however, not easy as it requires some logic so it works for everyone:

1. Check if any repo already uses https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Stable:/EL8-Uyuni-Client-Tools/EL_8/
2. Otherwise create a repository that uses https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Stable:/EL8-Uyuni-Client-Tools/EL_8/
3. Adjust all channels that use the repository with https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Stable:/CentOS8-
Uyuni-Client-Tools/CentOS_8/, to use the repository with https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Stable:/EL8-Uyuni-Client-Tools/EL_8/ instead.

Basically something similar to https://github.com/uyuni-project/uyuni/blob/master/schema/spacewalk/upgrade/susemanager-schema-4.3.3-to-susemanager-schema-4.3.4/002-use_almalinux_mirror.sql, but creating the required repository, if it's not there (for example for people managing CentOS8 only, but that never added Rocky or Alma).

For now, the workaround for users in this situation should be, before the upgrade:

1. At your Uyuni Server, via UI, find the repository using the URL https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Stable:/EL8-
Uyuni-Client-Tools/EL_8/
2. Then find the repository using the URL https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Stable:/CentOS8-Uyuni-Client-Tools/CentOS_8/
3. Find all channels using the repository from step 2, and adjust them to use the repository from step 1.
4. Remove repository from step 2
5. Apply the upgrade.

Ping @mcalmer, not sure if users can force a UI start if they already upgraded but the schema upgrade failed, so they can apply the workaround.

**Comment by EthanB11** [comment]
I ran into this and just commented out the SQL lines in the schema upgrade to get back into the system, because I forgot to grab a snapshot of the system after I made changes today before upgrading. I'd had two repo's previously one for Rocky Linux and one for CentOS/RedHat Linux. I checked both Uyuni Client tools repos and saw the Rocky Linux one had been updated to the new url I commented out. I then deleted the CentOS Uyuni tools repo and just linked that channel to the newer Rocky Linux Uyuni tools repo.

**Comment by phibid** [comment]
Sorry but I am trying to upgrade from 2021-09 to 2021-12, and I have this exact same issue it seems:

```
SQL>     select 'susemanager-schema-4.3.3-to-susemanager-schema-4.3.4/001-centos8-eol.sql.postgresql' from dual;
UPDATE rhnContentSource
SET source_url = 'https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/EL8-Uyuni-Client-Tools/EL_8/'
WHERE source_url = 'https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/CentOS8-Uyuni-Client-Tools/CentOS_8/' AND org_id IS NOT NULL;
UPDATE rhnContentSource
SET source_url = 'https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Stable:/EL8-Uyuni-Client-Tools/EL_8/'
WHERE source_url = 'https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Stable:/CentOS8-Uyuni-Client-Tools/CentOS_8/' AND org_id IS NOT NULL;
    commit;

                                      ?column?
-------------------------------------------------------------------------------------
 susemanager-schema-4.3.3-to-susemanager-schema-4.3.4/001-centos8-eol.sql.postgresql
(1 row)

UPDATE 0
psql:<stdin>:7: ERROR:  duplicate key value violates unique constraint "rhn_cs_repo_uq"
DETAIL:  Key (COALESCE(org_id, 0::numeric), type_id, source_url, (
CASE
    WHEN label::text ~~ 'manifest_%'::text THEN 1
    ELSE 0
END))=(1, 500, https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Stable:/EL8-Uyuni-Client-Tools/EL_8/, 0) already exists.
```

However strange thing is that I have not the client tools repo for CentOS 8, only the main one and AppStream, I have indeed the client tools repo for Alma 8, so I have removed it just in case of , but no luck, I can't still upgrade the Uyuni installation.

Any ideas ?

**Comment by juliogonzalez** [comment]
> Sorry but I am trying to upgrade from 2021-09 to 2021-12, and I have this exact same issue it seems:

Yes, 2021.12 is still broken if you have repos for CentOS8 and EL8 client tools. Could happen not only if you have CentOS, but also if you synced Oracle Linux 8 client tools in the past.

You should be able to proceed if you comment the lines at `susemanager-schema-4.3.3-to-susemanager-schema-4.3.4/001-centos8-eol.sql.postgresql` (can't recall the full path), or even better if you replace its content with the content of the SQL file from https://github.com/uyuni-project/uyuni/pull/4723, and then start the services again.

Needless to say: make sure you have a backup first.

**Comment by phibid** [comment]
Perfect thanks, applied the corrected version of /etc/sysconfig/rhn/schema-upgrade/susemanager-schema-4.3.3-to-susemanager-schema-4.3.4/001-centos8-eol.sql.postgresql and it works.

Philippe.


Philippe Bidault |  Unix Engineer |  Getronics
________________________________
M. 34617301667 |  E. ***@***.*** |  W. www.getronics.com<http://www.getronics.com>

Follow us on:

[cid:SocialLink_Linkedin_32x32_e7c75ee5-f8bf-4d58-979b-3050856d34e9.png]<https://www.linkedin.com/company/getronics/> [cid:SocialLink_Twitter_32x32_cb75fe15-abe1-46a3-9ac3-fc99d077e43b.png] <https://twitter.com/getronics>  [cid:SocialLink_Youtube_32x32_1946f212-29cc-40bc-9061-3536027a9813.png] <https://www.youtube.com/channel/UCfEFIFxLdxxhMN-Eaa6pXVA> <https://twitter.com/getronics>



Getronics CMC Service Desk Iberia S.L -  VAT No:S.L.: B66686262.
Registered Office - Getronics CMC Service Desk Iberia S.L, C/Rosselloi, Porcel, 21 planta 11, 08016 Barcelona, Spain.

The information transmitted is intended only for use by the addressee and may contain confidential and/or privileged material. Any review, re-transmission, dissemination or other use of it, or the taking of any action in reliance upon this information by persons and/or entities other than the intended recipient is prohibited. If you received this in error, please inform the sender and/or addressee immediately and delete the material.  Thank you.
Legal disclaimer: http://www.getronics.com/legal/ and further details of how we treat your personal data can be found in our privacy policy<https://www.getronics.com/policy-pages/corporate-policy-legal-disclaimer/>
